cli/src/diff_util.rs

```diff
@@ -398,6 +398,23 @@ fn show_color_words_diff_line(
     } else {
         write!(formatter, "    : ")?;
     }
+
+    // In color-words diffs, all hunks of DiffHunk::Different can be considered
+    // word-level and eligible for "token" highlighting. However, we exclude a
+    // hunk that spans the whole line because it would be too verbose if large
+    // block insertion or deletion were highlighted.
+    let write_diff_hunk = |formatter: &mut dyn Formatter, label: &str, data: &[u8]| {
+        if data.is_empty() {
+            return Ok(());
+        }
+        formatter.with_label(label, |formatter| {
+            if diff_line.has_left_content && diff_line.has_right_content {
+                formatter.with_label("token", |formatter| formatter.write_all(data))
+            } else {
+                formatter.write_all(data)
+            }
+        })
+    };
```

Comment:
This makes the display of full lines always match the display of, say, pluses and minuses in `jj diff --stat`. I think that's undesired.

We could add a "full_line" label to full lines. Perhaps we can think of a better solution?

Reply:
> This makes the display of full lines always match the display of, say, pluses and minuses in `jj diff --stat`.

Sorry, what do you mean by "this"? It's intentional that full-line hunks are rendered in the same way as `jj diff --git` (and `diff --git` will eventually gain word-level highlighting that looks similar to the current color-words diffs.)

> I think that's undesired. The `token` label (perhaps renamed) should exist for all the diff _text content_. In fact, one alternative name for it in my mind is `text`.

We can of course leave `token` labels for users who want to highlight all color-words hunks, but not git-diff hunks, but I'm not sure how many customization points we should support here. Something like `diff color_words added ..` might be better for that use case.

Reply:
> Sorry, what do you mean by "this"?

The labels "diff added" are used in more places than just `jj diff --git`. For example, they are used in `jj diff --stat`.

![image](https://github.com/martinvonz/jj/assets/4123047/9300cd47-5156-410a-9657-9db3de710401)

This was problematic before 848bb61.

So, I thought it would be unwise to again make that styling the same as the styling for text content. That's what I meant by "this". On second thought, maybe I stated things a bit strongly, and the specific solution I suggested isn't the best (see below). Also, perhaps this is a mild enough problem that it could wait for a separate PR.

> It's intentional that full-line hunks are rendered in the same way as jj diff --git

Makes sense. We might want to adjust the labels there too.

I'm not insisting on any particular solution, but one that is probably closer than what you had in mind would be to have both `jj diff` and `jj diff --git` label text with e.g. `diff text removed`. Then, using the `token` label just as you implemented it here would work fine (even if the `text` label appears in a separate PR).

Reply:
There was one more closely related thing I was thinking about why this matters. Let's say somebody dislikes this PR's change and wants to consistently underline all diff chunks. With this PR as is, unless I'm mistaken, they would have to add underlining the `diff added`/`diff removed` label. As a result, the pluses and minuses in `jj diff --stat` will be underlined too. Line numbers would also get underlined, additional config would be needed to explicitly disable underlining for `diff added line_number`.

But if the `text` label existed, they'd only underline `diff added text`/`diff removed text` and it would work. I also thought that `token` is meant to be `text`, but that was a separate confusion of mine.

Aside: another option is to use `chunk` instead of `text` (or instead of `token`?), judging by how I expressed my thoughts above.

Reply:
> Let's say somebody dislikes this PR's change and wants to consistently underline all diff chunks. With this PR as is, unless I'm mistaken, they would have to add underlining the `diff added`/`diff removed` label.

Correct. If we need to support both use cases (without changing the other diff styles), we'll need some extra labels. These labels might be `diff removed/added line`/`diff line removed/added token` in color-words diff, and `diff removed/added line token` in git diff, but I'm not sure.

I agree `token` isn't good name, but I can't find a better alternative. It's basically a "combined hunk of word/symbol-level diff". `text` has broader meaning, and `chunk` might be a bit confusing with `hunk`.

FWIW, word-level highlight in git diff will look like:
![image](https://github.com/martinvonz/jj/assets/172069/79a7e8f3-0e27-4afc-a1c4-1a2c55d9cbfc)

Reply:
> I agree token isn't good name, but I can't find a better alternative. It's basically a "combined hunk of word/symbol-level diff". 

It's indeed tricky.

It sounds slightly silly, but perhaps `mini-hunk` would work?

I opened up a thesaurus, and it suggested the nice word `morsel` as well as `block`, `bit`, `nugget`. Of these, the last is currently my favorite; a "diff nugget" sounds appealing to me.

Reply:
If we add "line" label that spans added/removed line, I think "word" is okay. I don't think the other suggestions are better than "token".

---

BTW, in order to copy unified-diff style from https://github.com/fowles/pratdiff, we'll need a different labeling structure. This PR assumes that user will highlight smaller hunks in addition to the default added/removed colors:
```xml
<removed>- <line><word>removed</word> word</line></removed>
<added>+ <line><word>added</word> word</line></added>
<added>+ <line>added line</line></added>
```
```toml
"removed" = "red"
"added" = "green"
"word" = { underline = true }
```

However, in pratdiff's style, it makes more sense to wrap all changed regions by `<word/>`:
```xml
<removed>- <line><word>removed</word> word</line></removed>
<added>+ <line><word>added</word> word</line></added>
<added>+ <line><word>added line</word></line></added>
```
```toml
"removed" = "red"
"added" = "green"
"removed line " = "gray"
"added line " = "default"
"removed line word" = "red"
"added line word" = "green"
```

In order to achieve both styles, it might be better to add something like "two_sided" for lines containing smaller hunks:
```xml
<removed>- <two_sided><line><word>removed</word> word</line></two_sided></removed>
<added>+ <two_sided><line><word>added</word> word</line></two_sided></added>
<added>+ <line><word>added line</word></line></added>
```
```toml
"removed" = "red"
"added" = "green"
"two_sided word" = { underline = true }
```

I think this is basically the same as the idea suggested in https://github.com/martinvonz/jj/pull/3958#discussion_r1657769643.

Another option is to use "changed"/"unchanged" labels like Mercurial:
```xml
<removed>- <line><changed>removed</changed><unchanged> word</unchanged></line></removed>
<added>+ <line><changed>added</changed><unchanged> word</unchanged></added>
<added>+ <line>added line</line></added>
```

This is backward compatible, but bad thing is that `"changed"` doesn't make sense in color-words diffs.